nemo/collections/asr/models/ctc_models.py

```diff
@@ -79,7 +80,28 @@ def setup_test_data(self, test_data_layer_params: Optional[Dict]):
         self.__test_dl = self.__setup_dataloader_from_config(config=test_data_layer_params)
 
     def setup_optimization(self, optim_params: Optional[Dict]):
-        self.__optimizer = torch.optim.Adam(self.parameters(), lr=optim_params['lr'])
+        optim_params = optim_params or {}  # In case null was passed as optim_params
+
+        # Check if caller provided optimizer name, default to Adam otherwise
+        optimizer_name = optim_params.get('optimizer', 'adam')
+
+        # Check if caller has optimizer kwargs, default to empty dictionary
+        optimizer_args = optim_params.get('opt_args', [])
+        optimizer_args = parse_optimizer_args(optimizer_args)
+
+        # We are guarenteed to have lr since it is required by the argparser
+        # But maybe user forgot to pass it to this function
+        lr = optim_params.get('lr', None)
+
+        if 'lr' is None:
+            raise ValueError('`lr` must be passed when setting up the optimization !')
+
+        # Actually instantiate the optimizer
+        optimizer = get_optimizer(optimizer_name)
+        self.__optimizer = optimizer(self.parameters(), lr=lr, **optimizer_args)
```

Comment:
why not merge these two lines into one?

Reply:
This wasn't what I had in mind actually. I was thinking more `return get_optimizer(optimizer_name, self.parameters(), lr=lr, **optimizer_args)`, ie I would expect get_optimizer to instantiate an optimizer for me.

If you want to keep your original design, I would actually prefer the old:
```
optimizer = get_optimizer(optimizer_name)
self.__optimizer = optimizer(self.parameters(), lr=lr, **optimizer_args)
```
rather than the changed:
```
optimizer = get_optimizer(optimizer_name)(self.parameters(), lr=lr, **optimizer_args)
```

Reply:
Oh I misunderstood. Yes, I'll revert to follow the older design. As to merging the two lines together, I would prefer not to do that for two reasons - 1) we may want the class without instantiation to wrap into another class (say we have experimental optimizer), 2) we want to pass the class as an argument without instantiation to perform defered computation or typecheck in tests.